contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,506 @@
+        IDepositSecurityModule dsm = IDepositSecurityModule(_locator.depositSecurityModule());
+        address[] memory guardians = dsm.getGuardians();
+        uint256 quorum = dsm.getGuardianQuorum();
+        dsm.addGuardians(guardians, quorum);
+    }
+
+    function _finishUpgrade() internal {
+        if (msg.sender != _voting) revert OnlyVotingCanUpgrade();
+        if (!isUpgradeStarted) revert StartMustBeCalledBeforeFinish();
+        if (isUpgradeFinished) revert CanOnlyFinishOnce();
+        /// Here we check that the contract got new ABI function getContractVersion(), although it is 0 yet
+        /// because in the new contract version is stored in a different slot
+        if (ILegacyOracle(_locator.legacyOracle()).getContractVersion() != 0) revert LidoOracleMustBeUpgradedToLegacy();
+        isUpgradeFinished = true;
+
+        ILegacyOracle(_locator.legacyOracle()).finalizeUpgrade_v4(_locator.accountingOracle());
+
+        ILido(_locator.lido()).finalizeUpgrade_v2(address(_locator), _eip712StETH);
+
+        INodeOperatorsRegistry(_nodeOperatorsRegistry).finalizeUpgrade_v2(
+            address(_locator),
+            _nodeOperatorsRegistryStakingModuleType,
+            _nodeOperatorsRegistryStuckPenaltyDelay
+        );
+
+        _grantRoles();
+
+        _passAdminRoleFromTemplateToVoting();
+
+        _verifyFinishedUpgrade();
+    }
+
+    function _prepareWithdrawalQueue() internal {
+        IOssifiableProxy(_locator.withdrawalQueue()).proxy__upgradeTo(_withdrawalQueueImplementation);
+        IWithdrawalQueue(_locator.withdrawalQueue()).initialize(address(this));
+        _resumePausableContract(_locator.withdrawalQueue());
+        _resumePausableContract(_locator.validatorsExitBusOracle());
+    }
+
+    function _prepareStakingRouter() internal {
+        IOssifiableProxy(_locator.stakingRouter()).proxy__upgradeTo(_stakingRouterImplementation);
+        IStakingRouter(_locator.stakingRouter()).initialize(
+            address(this),
+            _locator.lido(),
+            _withdrawalCredentials
+        );
+        _attachNORToStakingRouter();
+        // TODO: maybe attach NOR as module to staking router
+    }
+
+    function _attachNORToStakingRouter() internal {
+        bytes32 role = IStakingRouter(_locator.stakingRouter()).STAKING_MODULE_MANAGE_ROLE();
+        IAccessControlEnumerable(_locator.stakingRouter()).grantRole(role, address(this));
+        IStakingRouter(_locator.stakingRouter()).addStakingModule(
+            NOR_STAKING_MODULE_NAME,
+            _nodeOperatorsRegistry,
+            10000, // 100% target share
+            500, // 5% staking module fee
+            500 // 5% treasury fee
+        );
+        IAccessControlEnumerable(_locator.stakingRouter()).renounceRole(role, address(this));
+    }
+
+    function _grantRoles() internal {
+        IBurner burner = IBurner(_locator.burner());
+        burner.grantRole(burner.REQUEST_BURN_SHARES_ROLE(), _nodeOperatorsRegistry);
```

Comment:
It's better to check for non-upgradable contracts that these roles weren't polluted before the upgrade called

Reply:
done